### DISCUSSION TITLE
Ability to reset async atomWithDefault

### DISCUSSION THREAD

**Discussion by sazzer** [pmndrs/jotai#1460] [Q&A]
After [another question](https://github.com/pmndrs/jotai/discussions/1458) on here, I've now got an `atomWithDefault`, where the default is an async function to load data from my API. And that works great.

What I want to do now is to reset this atom when viewing a certain screen. In specific, the atom is the current user details, and when loading the user profile I want to re-fetch those details to make sure they're fresh.

And I can't work out how to do this. The obvious candidates haven't worked:
* useResetAtom
* Explicitly setting the atom to RESET
* Having another atom with a value of `new Date()` that I read in the getter for this function, and assign a new value in my reset call. Basically the atomWithRefresh pattern.

For reference I'm triggering the reset using `useEffect`. The reset function definitely gets called but the atom load function is *not* called - as I can see both from network calls but also from log statements.

Any other ideas what I can do here?

Cheers

**Comment by dai-shi** [comment]
https://jotai.org/docs/advanced-recipes/atom-creators#atom-with-refresh should do. I wonder what would be best practice in the future, but `atomWithRefresh` is a valid pattern.

**Reply by sazzer** [reply]
I've already tried this, and it works fine if the innermost atom is a normal `atom` but not if it's an `atomWithDefault`. 

Maybe this means the answer to my other question was wrong and `atomWithDefault` was the wrong thing to use there?

**Reply by sazzer** [reply]
This *mostly* works:
```
function asyncAtomWithStorage<T>(fn: (get: Getter) => T) {
  const storage = atom<T | Date>(new Date());

  return atom(
    (get) => {
      LOG('Reloading async atom');

      let stored = get(storage);
      if (stored instanceof Date) {
        stored = fn(get);
      }
      return stored;
    },
    (get, set, update: T) => {
      LOG('Updating async atom');

      return set(storage, update === RESET ? new Date() : update);
    }
  );
}
```

I say "mostly" because I can't quite work out how to get the TS type definitions correct, because when I'm calling `setUser(RESET)` it's complaining that "Argument of type 'typeof RESET' is not assignable to parameter of type 'Promise<State<UserProperties>>'.ts(2345)"

I suspect it's a relatively easy tweak if I can work it out though :)

**Reply by dai-shi** [reply]
If you combine atomWithDefault and atomWithRefresh, you would probably need to RESET when updating the refresh counter. 

Your solution seems better and simpler. Probably, you just need to change the type of update to `T | typeof RESET`.

**Reply by sazzer** [reply]
That worked, in conjunction with making the inner atom be `atom<T | Date | typeof RESET>` as well :)

**Reply by RyanThomas73** [reply]
@dai-shi 
Is there a mechanism/pattern by which we can trigger a refresh when using the `loadable` util?

Looking at the code it looks like there is an internally maintained refresh atom (for debugging?) but the end result wraps the `derivedAtom` in a read-only atom so the setter to trigger the refresh isn't exposed.

And trying the atomWithRefresh pattern with the underlying atom would require a useAtom dependency on the underlying atom which would still suspend...

**Reply by dai-shi** [reply]
> And trying the atomWithRefresh pattern with the underlying atom would require a useAtom dependency on the underlying atom which would still suspend...

Can you elaborate? I thought, the atomWithRefresh pattern works with `loadable`.
